File: api/_src/modules/identity/authorization.ts
/**
 * Deny-by-default role authorization primitives
 * (port of platform_app.modules.identity.authorization).
 */
import type { RoleName } from "@energy/shared";

export class AccessDeniedError extends Error {
  constructor(message = "operation is not authorized") {
    super(message);
    this.name = "AccessDeniedError";
  }
}

export const Permission = {
  VIEW_SELF: "view_self",
  VIEW_BUILDING: "view_building",
  MANAGE_BUILDING: "manage_building",
  SELECT_TENANT: "select_tenant",
} as const;

export type Permission = (typeof Permission)[keyof typeof Permission];

export interface RoleGrant {
  role: RoleName;
  universityId: string | null;
  buildingId: string | null;
}

export interface Principal {
  accountId: string;
  grants: RoleGrant[];
}

export interface RequireOptions {
  universityId: string;
  buildingId?: string | null;
  platformTargetExplicit?: boolean;
}

/** Authorize only explicit role, tenant, and building combinations. */
export class AuthorizationService {
  require(
    principal: Principal,
    permission: Permission,
    { universityId, buildingId = null, platformTargetExplicit = false }: RequireOptions,
  ): void {
    for (const grant of principal.grants) {
      if (grant.role === "platform_admin") {
        if (platformTargetExplicit) {
          return;
        }
        continue;
      }
      if (grant.universityId !== universityId) {
        continue;
      }
      if (grant.role === "participant" && permission === Permission.VIEW_SELF) {
        return;
      }
      if (grant.role === "building_admin") {
        if (permission === Permission.VIEW_SELF) {
          return;
        }
        if (
          (permission === Permission.VIEW_BUILDING ||
            permission === Permission.MANAGE_BUILDING) &&
          buildingId !== null &&
          grant.buildingId === buildingId
        ) {
          return;
        }
      }
    }
    throw new AccessDeniedError();
  }
}

File: src/platform_app/modules/identity/authorization.py
"""Deny-by-default role authorization primitives."""

from dataclasses import dataclass
from enum import StrEnum
from uuid import UUID

from platform_app.modules.identity.models import Role


class AccessDeniedError(PermissionError):
    """Raised when no explicit role grant permits an operation."""


class Permission(StrEnum):
    VIEW_SELF = "view_self"
    VIEW_BUILDING = "view_building"
    MANAGE_BUILDING = "manage_building"
    SELECT_TENANT = "select_tenant"


@dataclass(frozen=True, slots=True)
class RoleGrant:
    role: Role
    university_id: UUID | None
    building_id: UUID | None


@dataclass(frozen=True, slots=True)
class Principal:
    account_id: UUID
    grants: tuple[RoleGrant, ...]


class AuthorizationService:
    """Authorize only explicit role, tenant, and building combinations."""

    def require(
        self,
        principal: Principal,
        permission: Permission,
        *,
        university_id: UUID,
        building_id: UUID | None = None,
        platform_target_explicit: bool = False,
    ) -> None:
        for grant in principal.grants:
            if grant.role is Role.PLATFORM_ADMIN:
                if platform_target_explicit:
                    return
                continue
            if grant.university_id != university_id:
                continue
            if grant.role is Role.PARTICIPANT and permission is Permission.VIEW_SELF:
                return
            if grant.role is Role.BUILDING_ADMIN:
                if permission is Permission.VIEW_SELF:
                    return
                if (
                    permission in {Permission.VIEW_BUILDING, Permission.MANAGE_BUILDING}
                    and building_id is not None
                    and grant.building_id == building_id
                ):
                    return
        raise AccessDeniedError("operation is not authorized")
